Add Power Management support to TI cc23x0 SoC

bogdanovs opened the pull request:
This series adds Power Management support to TI cc23x0 SoC.

Datasheet: https://www.ti.com/lit/ds/symlink/cc2340r5.pdf

bogdanovs committed:
soc: ti: cc23x0: Add power management

Add power management capabilities for cc23x0:
- runtime-idle
- standby
- soft-off

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

jpanisbl committed:
dts: bindings: clock: Add TI cc23x0 external low-frequency oscillator

External 32.768 kHz crystal oscillator (LFXT) connected across
the X32P input and X32N output pins.

Signed-off-by: Julien Panis <jpanis@baylibre.com>

jpanisbl committed:
boards: ti: lp_em_cc2340r5: Add external low-frequency oscillator

Enable external 32 kHz crystal oscillator.

Signed-off-by: Julien Panis <jpanis@baylibre.com>

bogdanovs committed:
dts: arm: ti: cc23x0: Add power management support

Add support for PM to cc23x0 SoC.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>
Signed-off-by: Julien Panis <jpanis@baylibre.com>

bogdanovs committed:
drivers: timer: Add RTC timer driver for cc23x0

Add support for RTC as timer for cc23x0 SoC.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

bogdanovs committed:
drivers: timer: cc23x0: Add option to select between RTC and SYSTIM

Add choice menu where could be selected between RTC or SYSTIM for
system timer.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

bogdanovs committed:
drivers: counter: cc23x0: Add dependency for RTC

In case RTC is used for system timer, it should not be used
as counter device. Dependency restricts counter driver to work
only with SYSTIM.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

bogdanovs committed:
soc: ti: cc23x0: Add clock definition for RTC

Add conditonal definition for RTC and SYSTIM with different
values for both of them respecting clock speed and ticks
per minute.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

bogdanovs committed:
soc: ti: cc23x0: Add conditions for RTC as timer in power.c

In power management, add conditions to handle the case where
RTC is used as main timer.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

bogdanovs committed:
soc: ti: cc23x0: Add support for RTC alarms in power.c

In power management, add support to take into account the alarms
set in RTC. Alarm from RTC is processed like any other from SYSTIM.
This prevents from missing interrupts.

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

bogdanovs commented:
Hi , @andyross @nordic-krch @teburd @vaishnavachath @vanti can you review when you have time ?

gumulka reviewed (commented):
Will have a deeper look later, but some preliminary comments.

gumulka commented on drivers/timer/cc23x0_rtc_timer.c:
```diff
@@ -0,0 +1,122 @@
+/* Keep track of rtc counter at previous announcement to the kernel */
+static uint32_t last_rtc_count;
+
+static void rtc_isr(const void *arg);
+static int sys_clock_driver_init(void);
```
No need to define these functions. They are defined before they are needed later.

gumulka commented on drivers/timer/cc23x0_rtc_timer.c:
```diff
@@ -0,0 +1,122 @@
+
+uint32_t sys_clock_elapsed(void)
+{
+	uint32_t current_rtc_count = HWREG(RTC_BASE + RTC_O_TIME8U);
+
+	uint32_t elapsed_rtc;
+
+	if (current_rtc_count >= last_rtc_count) {
+		elapsed_rtc = current_rtc_count - last_rtc_count;
+	} else {
+		elapsed_rtc = (0xFFFFFFFF - last_rtc_count) + current_rtc_count;
+	}
+
```
This is the same as the start of rtc_isr. Could you create a static function, that does that and remove the code duplicate?

gumulka commented on drivers/timer/cc23x0_rtc_timer.c:
```diff
@@ -0,0 +1,122 @@
+		uint32_t ticks_now = HWREG(RTC_BASE + RTC_O_TIME8U);
+
+		if ((ticks_now + ticks) >= RTC_TIMEOUT_MAX) {
+			/* Reset timer and start from 0 */
+			HWREG(RTC_BASE + RTC_O_CTL) = 0x1;
+			HWREG(RTC_BASE + RTC_O_CH0CC8U) = ticks;
```
I did not quite understand this, but could it be, that this will make the calculation of the elapsed tick inaccurate, because the timer was reset and the last_rtc_count was not?

gumulka commented on drivers/timer/cc23x0_rtc_timer.c:
```diff
@@ -0,0 +1,122 @@
+	now_ticks = HWREG(RTC_BASE + RTC_O_TIME8U);
+	last_rtc_count = now_ticks;
+
+	HWREG(RTC_BASE + RTC_O_ICLR) = 0x3;
+	HWREG(RTC_BASE + RTC_O_IMCLR) = 0x3;
```
Is there a reason for this magic number?

gumulka commented on soc/ti/simplelink/cc23x0/Kconfig.defconfig:
```diff
@@ -10,8 +10,20 @@ if SOC_SERIES_CC23X0
+		default 32768
+
+	config SYS_CLOCK_TICKS_PER_SEC
+		default 125375
```
Where does this value come from? It seems unusually large.

bogdanovs replied:
It was calculated , it is not a magic number. It is result from ticks which sys clock is making calculated to correct resolution

gumulka replied:
Could you share how it was calculated in the commit message or as a comment?

gumulka commented on soc/ti/simplelink/cc23x0/power.c:
```diff
@@ -0,0 +1,245 @@
+	case PM_STATE_STANDBY:
+		pm_cc23x0_enter_standby();
+		break;
+	case PM_STATE_SOFT_OFF:
+		Power_shutdown(0, 0);
+		break;
```
Since this is shutdown, wouldn't it make more sense to implement z_sys_poweroff instead of having it as a power mode?

Because, after going into this state, as far as I understand the datasheet, you wake up by going through a whole boot cycle. So shutdown seems more appropriate from my side.

gumulka replied:
I have the assumption, that the function `Power_shutdown(0, 0);` will shut down the CPU. Meaning, it will not resume from the shutdown in this function with the RAM content intact, but instead do a full boot cycle afterwards if an interrupt wakes it up.

If my assumption holds up, then this is not part of power-management. Power management should the the automatic power management, that saves power by going to sleep and resume execution as if nothing happened in between.

There is another API for powering down the CPU, which is to implement the `z_sys_poweroff` function for the CPU. If you do "git grep z_sys_poweroff" you will see many [examples](https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/ti/mspm0/common/poweroff.c) for it.

gumulka replied:
No, you are right, I was not aware of that. It is now just that I don't like that behaviour and there is not much in the codebase, that does it that way. Most go for the z_sys_poweroff implementation, which I find cleaner.

bogdanovs committed:
drivers: serial: cc23x0: Replace incorrect macros

Replace CONFIG_PM with CONFIG_PM_DEVICE to include pm_lock
in cc23x0 UART driver in struct uart_cc23x0_data

Signed-off-by: Stoyan Bogdanov <sbogdanov@baylibre.com>

bogdanovs commented on soc/ti/simplelink/cc23x0/power.c:
```diff
@@ -0,0 +1,245 @@
+	ARG_UNUSED(substate_id);
+
+	switch (state) {
+	case PM_STATE_RUNTIME_IDLE:
+		PowerCC23X0_doWFI();
+		break;
```
yes , I keep it that way for consistency. if you insist it is confusing It could be replaced with `__WFI`.

bogdanovs commented on boards/ti/lp_em_cc2340r5/lp_em_cc2340r5.dts:
```diff
@@ -62,6 +62,10 @@
 		};
 	};
+
+	lfxosc: lf_xosc {
+		compatible = "ti,cc23x0-lf-xosc";
+	};
```
This stand for Low Frequency External Oscillator which hardware component which could or could not be populated on different board with cc2340r5 SoCs.

```
External oscillator (LFXT): external 32.768 kHz crystal oscillator connected across the X32P input and
X32N output pins.
```

kartben commented:
re-assigning to TI SimpleLink maintainer.
@bjarki-andreasen  maybe you can help review too? Thanks!
